test/query-builder.js

```diff
@@ -0,0 +1,131 @@
+import test from 'tape'
+import QueryBuilder from '../src/query-builder'
+
+test('QueryBuilder should build query with no field', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('match_all')
+
+  t.deepEqual(result._queries, {
+    match_all: {}
+  })
+})
+
+test('QueryBuilder should build query with field but no value', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('exists', 'user')
+
+  t.deepEqual(result._queries, {
+    exists: {
+      field: 'user'
+    }
+  })
+})
+
+test('QueryBuilder should build query with field and value', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('term', 'user', 'kimchy')
+
+  t.deepEqual(result._queries, {
+    term: {
+      user: 'kimchy'
+    }
+  })
+})
+
+test('QueryBuilder should build query with field and object value', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('range', 'date', {gt: 'now-1d'})
+
+  t.deepEqual(result._queries, {
+    range: {
+      date: {gt: 'now-1d'}
+    }
+  })
+})
+
+test('QueryBuilder should build query with more options', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('geo_distance', 'point', {lat: 40, lon: 20}, {distance: '12km'})
+
+  t.deepEqual(result._queries, {
+    geo_distance: {
+      distance: '12km',
+      point: {
+        lat: 40,
+        lon: 20
+      }
+    }
+  })
+})
+
+test('QueryBuilder should build nested queries', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('nested', 'path', 'obj1', (q) => q.query('match', 'obj1.color', 'blue'))
+
+  t.deepEqual(result._queries, {
+    nested: {
+      path: 'obj1',
+      query: {
+        match: {
+          'obj1.color': 'blue'
+        }
+      }
+    }
+  })
+})
+
+test('QueryBuilder should nest bool-merged queries', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().query('nested', 'path', 'obj1', {score_mode: 'avg'}, (q) => {
+    return q.query('match', 'obj1.name', 'blue').query('range', 'obj1.count', {gt: 5})
+  })
+
+  t.deepEqual(result._queries, {
+    nested: {
+      path: 'obj1',
+      score_mode: 'avg',
+      query: {
+        bool: {
+          must: [
+            {
+              match: {'obj1.name': 'blue'}
+            },
+            {
+              range: {'obj1.count': {gt: 5}}
+            }
+          ]
+        }
+      }
+    }
+  })
+})
+
+test('QueryBuilder should make filter aggregations', (t) => {
+  t.plan(1)
+
+  const result = new QueryBuilder().aggregation('filter', null, 'red_products', (b) => {
+    return b.filter('term', 'color', 'red').aggregation('avg', 'price', 'avg_price')
```

Comment:
My thought was to be as flexible as possible, so this approach allows combining queries, filters and aggregations arbitrarily. Here are a few more examples where this type of composition is needed: 
- [constant score query, filter in a query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-constant-score-query.html).
- [bool query, filter in a query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html)

Is your concern that this approach opens the possibility for someone to build an invalid query?